apps/typegpu-docs/src/content/docs/fundamentals/slots.mdx

```diff
@@ -1,63 +1,168 @@
+- `.$` shorthand property.
+
+```typescript
+const value = mySlot.value; // or mySlot.$
+```
 
+:::caution
+Slot values can only be accessed inside TypeGPU functions.
+Attempting to access them outside will throw an error.
+:::
+
+Slots are resolved during the TypeGPU resolution phase.
+The slot object itself does not hold any value.
+Instead, you can bind it in one of two ways.
+
+### Binding after defining a function
+
+The first way to bind a value to a slot is to call the `with` method on a `TgpuFn` instance:
+
+```ts twoslash
+import tgpu from 'typegpu';
+import * as d from 'typegpu/data';
+import { mul } from 'typegpu/std';
+// ---cut---
+const filterColorSlot = tgpu.slot<d.v3f>();
+
+const filter = tgpu.fn([d.vec3f], d.vec3f)((color) => {
+  const filterColor = filterColorSlot.$;
   return mul(color, filterColor);
 });
 
 // Bind the filter function with a red color.
 const filterWithRed = filter
-  .with(filterColorSlot, vec3f(1, 0, 0));
+  .with(filterColorSlot, d.vec3f(1, 0, 0));
 
 // Bind the filter function with a green color.
 const filterWithGreen = filter
-  .with(filterColorSlot, vec3f(0, 1, 0));
+  .with(filterColorSlot, d.vec3f(0, 1, 0));
 ```
 
-### Conditional compilation
+In the example above, after resolution we are left with two different WGSL functions:
 
-```ts
-const safeLoopSlot = tgpu.slot(false);
+```wgsl
+fn filter_0(color: vec3f) ->  vec3f{
+  var filterColor = vec3f(1, 0, 0);
+  return (color * filterColor);
+}
 
-const processObjects = tgpu.fn([f32], f32)(() => {
-  let len = objects.length;
+fn filter_1(color: vec3f) ->  vec3f{
+  var filterColor = vec3f(0, 1, 0);
+  return (color * filterColor);
+}
+```
 
-  if (safeLoopSlot.$) { // <- evaluated at compile time
-    // This will be included in the shader code only if
-    // `safeLoopSlot` is bound to true in the caller's scope.
-    len = max(len, 9999);
-  }
+### Binding during pipeline creation
 
-  for (let i = 0; i < len; ++i) {
-    // ...
-  }
-});
+The other way to fill in a slot is to call the `with` method during the creation of a `TgpuComputePipeline` or `TgpuRenderPipeline`:
+
+```ts twoslash
+import tgpu from 'typegpu';
+import * as d from 'typegpu/data';
+
+const root = await tgpu.init();
+// ---cut---
+const resultBuffer = root.createMutable(d.i32, 0);
+const multiplierSlot = tgpu.slot<number>();
+
+const computeMultiply = tgpu['~unstable']
+  .computeFn({ workgroupSize: [1] })(() => {
+    resultBuffer.$ = resultBuffer.$ * multiplierSlot.$;
+  });
+
+const pipeline = root['~unstable']
+  .with(multiplierSlot, 3)
+  .withCompute(computeMultiply)
+  .createPipeline();
 ```
 
-### Inversion of control
+The pipeline above resolves to the following WGSL:
 
-```ts
+```wgsl
+@group(0) @binding(0) var<storage, read_write> resultBuffer_1: i32;
+
+@compute @workgroup_size(1) fn computeMultiply_0(){
+  resultBuffer_1 = (resultBuffer_1 * 3);
+}
+```
+
+:::note
+Static code analysis does not incorporate whether all slots were filled or not.
```

Comment:
What does incorporate mean in this context? Maybe there's a way to rephrase this to be using simpler wording?